Mhd/layer contents integration

mhdostal opened the pull request:
Included in this PR:

- LayerContents component based on [this](https://github.com/ArcGIS/open-source-apps/blob/master/app/data-collection/components/layer-contents/Spec.md) design.  Note that the title and title icon for both the Layers and Bookmarks view is under design review and not included here.

- Inclusion of an "Extras" button in the right side of the app's title bar which displays an AlertViewController allowing the user to select either Layers or Bookmarks.

- Integration of the LayerContents component in DataCollection.

- Integration of the Bookmarks component from the Toolkit (and inclusion of the Toolkit as a submodule).

Extras alert:
![Extras](https://user-images.githubusercontent.com/3998072/79911136-7ad8cf00-83e5-11ea-8dbc-e958e849aca8.png)

Layers:
![Layers](https://user-images.githubusercontent.com/3998072/79911139-7c09fc00-83e5-11ea-8b6f-6c5e0503d321.png)

Bookmarks:
![Bookmarks](https://user-images.githubusercontent.com/3998072/79911143-7ca29280-83e5-11ea-8d31-2803284c2bc7.png)

philium commented on data-collection/data-collection/View Controllers/Map View Controller/MapViewController+Extras.swift:
```diff
@@ -0,0 +1,113 @@
+        let action = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
+                
+        // Present the list of extras.
+        for extra in [layersExtra, bookmarksExtra] {
+            let extraAction = UIAlertAction(title: extra, style: .default, handler: { [weak self] (action) in
+                switch extra {
+                case layersExtra:
+                    self?.showLayerContents(barButtonItem)
+                case bookmarksExtra:
+                    self?.showBookmarks(barButtonItem)
+                default:
+                    break
+                }
+            })
+            
+            var image: UIImage?
+            switch extra {
+            case layersExtra:
+                image = #imageLiteral(resourceName: "legend")
+            case bookmarksExtra:
+                image = #imageLiteral(resourceName: "bookmark")
+            default:
+                break
+            }
+            
+            extraAction.setValue(image, forKey: "image")
+            action.addAction(extraAction)
+        }
```
It seemed like the `for` loop was a convenience…until I saw the two switch statements. Now I'm wondering whether the code would be simpler without the for loop.

mhdostal replied:
rewrote this bit, please take another look.

philium commented on data-collection/data-collection/View Controllers/Map View Controller/MapViewController+Extras.swift:
```diff
@@ -0,0 +1,113 @@
+                break
+            }
+            
+            extraAction.setValue(image, forKey: "image")
```
I really don't recommend doing this. I know that it works now, but there's no guarantee it will continue to work in the future. As much as I would love for Apple to expose a way to set an image on an alert action, they haven't, and I can only assume there's a good reason for that. Apple is [pretty clear](https://developer.apple.com/documentation/uikit/windows_and_screens/getting_the_user_s_attention_with_alerts_and_action_sheets) about what alerts and action sheets should be used for ("important information…important choice"). I've been guilty of violating that and using an action sheet out of convenience, but I don't recommend it. If at all possible, implement your own view for this.

mhdostal replied:
I'm surprised it's not supported because it does add some flair to an otherwise unremarkable interface element.  I know I've seen it in other things we've done, but I can't seem to find that now, so maybe it was deleted there as well.  I'll delete them.

philium commented on data-collection/data-collection/View Controllers/Map View Controller/MapViewController+Extras.swift:
```diff
@@ -0,0 +1,113 @@
+    
+    func showLayerContents(_ barButtonItem: UIBarButtonItem?) {
+        // Create the LayerContentsViewController if it doesn't exist.
+        if layerContentsViewController == nil {
+            let dataSource = DataSource(geoView: mapView)
+            layerContentsViewController = TableOfContents(dataSource)
+            
+            // Add a done button.
+            let doneButton = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(done))
+            layerContentsViewController?.navigationItem.leftBarButtonItem = doneButton
+        }
+        
+        if let layerContentsVC = layerContentsViewController {
```
You could make the second `if` statement unnecessary by using an `if…else` statement:

```swift
let layerContentsVC: LayerContentsViewController
if let existingViewController = layerContentsViewController {
    layerContentsVC = existingViewController
} else {
    // create and configure the view controller
    layerContentsVC = ...
}
```

Same in `showBookmarks(_:)`.

mhdostal commented on data-collection/data-collection/View Controllers/App Container/Base.lproj/AppContainerViewController.storyboard:
```diff
@@ -85,6 +83,11 @@
                             </connections>
                         </barButtonItem>
                         <rightBarButtonItems>
+                            <barButtonItem title="Item" image="ellipsis" catalog="system" id="4IE-fC-WYU" userLabel="Extras Item">
```
That is odd that it's the default;  I made the change to `plain` on all of the bar button items.

mhdostal replied:
I reverted this change because I got the following warning after changing them all to `plain`.

![image](https://user-images.githubusercontent.com/3998072/80002926-e96d6980-8485-11ea-96e4-4db63570208d.png)

philium replied:
Fascinating! That isn't at all confusing. Good to know!

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/DataSource.swift:
```diff
@@ -0,0 +1,126 @@
+    /// Returns a `DataSource` initialized with the given `AGSLayerContent` array..
+    /// - Parameter layers: The array of `AGSLayerContent`.
+    /// - Since: 100.8.0
+    public init(layers: [AGSLayerContent]) {
```
I'm cool with that.  @esreli - You OK with this suggestion?

esreli replied:
Yes!

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/DataSource.swift:
```diff
@@ -0,0 +1,126 @@
+/// or an `AGSGeoView` from whose `AGSMap` or `AGSScene` the  operational and
+/// base map layers (`AGSLayerContent`) are extracted.
+/// - Since: 100.8.0
+public class DataSource: NSObject {
```
We did at one point, but not sure why we didn't make the change;  @esreli ?

esreli replied:
I think the reason it did not change was that I had thought it would serve well as a member of `LayerContentsTableViewController`. For example:

```swift
class LayerContentsTableViewController {
    class DataSource { 
        /* ... */ 
    }
}
```

mhdostal replied:
@philium @esreli 
I prefer it as it's own class, but I'm not averse to changing the name.  Does `LayerContentsDataSource` make sense?

esreli replied:
Yes it does.

On the side... after learning more about the relationship between `LayerContentsTableViewController` and `LayerContentsViewController`, I would correct my above suggestion to `LayerContentsViewController`. Of course, this suggestion is moot as its own class.

mhdostal replied:
Renamed to `LayerContentsDataSource`

mhdostal commented on data-collection/data-collection/View Controllers/App Container/Base.lproj/AppContainerViewController.storyboard:
```diff
@@ -85,6 +83,11 @@
                             </connections>
                         </barButtonItem>
                         <rightBarButtonItems>
+                            <barButtonItem title="Item" image="ellipsis" catalog="system" id="4IE-fC-WYU" userLabel="Extras Item">
```
An oversight on my part; I was going to change it but just forgot about it.  Thanks!  I changed it to the Calcite Ellipsis.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/DataSource.swift:
```diff
@@ -0,0 +1,126 @@
+        }
+    }
+    
+    private func appendBasemap(_ basemap: AGSBasemap?) {
+        guard let basemap = basemap else { return }
+        
```
I made the change and removed an unnecessary `let mapView = ...`.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.storyboard:
```diff
@@ -0,0 +1,199 @@
+                                    <outlet property="visibilitySwitch" destination="Qjr-f6-CC9" id="49i-PM-IbO"/>
+                                </connections>
+                            </tableViewCell>
+                            <tableViewCell clipsSubviews="YES" contentMode="scaleToFill" selectionStyle="default" indentationWidth="10" reuseIdentifier="SublayerTitle" id="bWm-du-xG7" customClass="LayerCell" customModule="LayerContentsForiPad" customModuleProvider="target">
```
The `Name` label uses different text in the `LayerTitle` vs. `SublayerTitle` cells.  The functionality is identical, but I wanted two different cells for customization purposes.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.swift:
```diff
@@ -0,0 +1,231 @@
+let indentationConstant: CGFloat = 10.0
+
+/// The protocol you implement to respond to user content accordion changes.
+internal protocol ContentAccordionDelegate: AnyObject {
```
Good suggestion.  Done.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.swift:
```diff
@@ -0,0 +1,231 @@
+        }
+    }
+    
+    var layerIndentationLevel: Int = 0 {
```
Yes, and it didn't work for me for some reason.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.storyboard:
```diff
@@ -0,0 +1,199 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
- I'm getting two constraint warnings.  The first has to do with displaying the `UIAlertController` as an `.actionSheet`.  I think that one is a bug on Apple's end.  I get the same message when doing the same thing in the AR Toolkit and it is present in the iOS Samples app as well (in the `DownloadPreplannedMapViewController.removeDownloadsTapped` method).  I found [this](https://stackoverflow.com/questions/55653187/swift-default-alertviewcontroller-breaking-constraints) StackOverflow post.

The second one is: `Warning once only: Detected a case where constraints ambiguously suggest a height of zero for a table view cell's content view.`  I suspect it's because the cell is being displayed before the data is loaded, but I haven't looked into it.

- The Done/Cancel button thing is something that is under design review.

- The rows shouldn't be selectable; I'll fix that.

- fixed the cell spacing.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.swift:
```diff
@@ -0,0 +1,231 @@
+/// LayerCell - cell representing either a Layer or LayerContent (i.e., a sublayer).
+/// The difference is in which table view cell prototype is used for the cell.
+class LayerCell: UITableViewCell {
+    var content: Content? {
```
I removed the `contents` property from the TableViewCells and put all that logic into the TableViewDataSource methods.  I needed to redo the LayerCellDelegate as well, but I think it's better for it.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.swift:
```diff
@@ -0,0 +1,231 @@
+        let newAccordian: AccordionDisplay = (content?.accordion == .expanded) ? .collapsed : .expanded
+        content?.accordion = newAccordian
+        button.setImage(LayerContentsTableViewController.accordianImage(newAccordian), for: .normal)
+        accordionDelegate?.accordionChanged()
```
I would prefer to only update what changed, but the issue with the accordion functionality is that expanding/contracting the accordion can add or remove several rows in the table.  The `visibleContent` array is what's currently shown.  That is created from the `contents` array based on whether an item's parents.accordion is collapsed or not.  I could do a diff before and after the change, but then I'd have to match the `contents` to the TableViewCell/indexPath and update the table with that.  If I get a chance I'll look into it.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,474 @@
+        contents.removeAll()
+        displayedLayers.removeAll()
+        
+        guard let layerContents = dataSource?.layerContents else { layerContentsTableViewController?.contents = [Content](); return }
```
Rewritten to be more understandable.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.swift:
```diff
@@ -0,0 +1,231 @@
+                        // wasn't already recycled by the time we get the swatch.
+                        if cell.tag != indexPath.hashValue {
+                            return
+                        }
```
I used the `symbol.hashValue` instead of the `indexPath.hashValue`.  That way we know that the cell in question contains the symbol the image is retrieved from.  And I'm getting the `indexPath` for the cell and just updating that row.

mhdostal commented:
@philium - I'm done making the suggested changes and addressing comments, so please give it another look.

The outstanding issues:

- any potential renaming/moving of the DataSource class
- LayerContentsTableViewController issues:  Done/Cancel button and TableViewCell constraint issue
- Find a strategy to eliminate unnecessary loading of tableView data.

Thanks again for your review, it really helps make the code better.

philium commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,480 @@
+
+/// Configuration is an protocol (interface) that drives how to format the layer contents table.
+/// - Since: 100.8.0
+public protocol LayerContentsConfiguration {
```
Unless I'm missing something, adopting types are identical, which makes me think this might make more sense as a struct than a protocol. Then `TableOfContents` and `Legend` could just be static members of the type:

```swift
struct LayerContentsConfiguration { ... }

extension LayerContentsConfiguration {
    /// The configuration for a table of contents.
    static let tableOfContents = ...
    /// The configuration for a legend.
    static let legend = ...
}
```

mhdostal replied:
@philium Unless you have strong objections, because of the desire to have actual classes for the Legend and TableOfContents, I'd prefer to leave this the way it is.

philium commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,480 @@
+
+/// Describes a `LayerContents` view configured to display a legend.
+/// - Since: 100.8.0
+public class Legend: LayerContentsViewController {
```
Could this be named `LegendViewController`?

mhdostal replied:
I'm Ok with that.  @esreli ?

esreli replied:
I like it!

philium commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,480 @@
+
+/// Describes a `LayerContents` view configured to display a table of contents.
+/// - Since: 100.8.0
+public class TableOfContents: LayerContentsViewController {
```
Could this be named `TableOfContentsViewController`?

That being said, it seems like these subclasses aren't even needed. Incorporating my previous suggestion, wouldn't the following be equivalent?

```swift
LayerContentsViewController(dataSource: dataSource, config: .legend) // Legend(dataSource: dataSource)
LayerContentsViewController(dataSource: dataSource, config: .tableOfContents) // TableOfContents(dataSource: dataSource)
```

mhdostal replied:
I'm Ok with the renaming. @esreli ?

The reason they are separate subclasses is from the Toolkit Design Council.  They thought the API for these two common scenarios (Legend and TOC) would be more easily understood if there were classes for both.

esreli replied:
@philium the subclasses were requested by Toolkit stakeholders. They were under the impression that our customers are accustomed to a **Legend** and a **TableOfContents** but not a **LayerContents**. While I agree with you here, because it's not hard to formalize these tools as subclasses, I say it's value added to make these types concrete.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.swift:
```diff
@@ -0,0 +1,220 @@
+                    symbol.createSwatch(completion: { [weak self] (image, _) -> Void in
+                        // Make sure this is the cell we still care about and that it
+                        // wasn't already recycled by the time we get the swatch.
+                        if cell.tag != symbol.hashValue {
+                            return
+                        }
+                        
+                        // Set the swatch into our dictionary and reload the row
+                        self?.symbolSwatches[symbol] = image
```
Good point.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.swift:
```diff
@@ -0,0 +1,220 @@
+                    // We have a swatch, so set it into the imageView.
+                    legendInfoCell.symbolImage = swatch
+                } else {
+                    // Tag the cell so we know what symbol is being used
+                    // to create the swatch.  This will ensure we
+                    // put the legend image in the correct cell.
+                    cell.tag = symbol.hashValue
```
Ok, got it.

esreli commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,480 @@
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Can we consider supporting Storyboards?

esreli commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,480 @@
+    private var contents = [Content]()
+    
+    // Array of child:parent used when determining accordion status.
+    private var parents = [UInt : [AGSLayerContent]]()
```
I haven't looked into this too deeply yet, though I always am skeptical of a child maintaining a reference to its parent. Is this potentially introducing a retain cycle?

mhdostal replied:
The relationship is only one way; child contain references to their parents, but the parents do not contain references to their children, so there should be no retain cycle.

esreli commented on data-collection/data-collection/View Controllers/Map View Controller/MapViewController+Extras.swift:
```diff
@@ -0,0 +1,126 @@
+import ArcGIS
+import ArcGISToolkit
+
+private let layersExtra = "Layers"
```
These global variables can probably be computed from within their `enum Extras` counterpart. I would create another extension on `Extras` that supplies these titles.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/DataSource.swift:
```diff
@@ -0,0 +1,122 @@
+    public init(geoView: AGSGeoView) {
+        super.init()
+        self.geoView = geoView
+        geoViewDidChange()
```
I checked and it does *not* trigger the `didSet` when setting the `geoView` in `init`.

philium commented on data-collection/data-collection/View Controllers/Map View Controller/MapViewController+Extras.swift:
```diff
@@ -0,0 +1,126 @@
+                
+        // Present the list of extras.
+        for extra in Extras.allCases {
+            var extraAction: UIAlertAction
```
This doesn't need to be mutable.

```suggestion
            let extraAction: UIAlertAction
```

philium commented on data-collection/data-collection/View Controllers/Map View Controller/MapViewController+Extras.swift:
```diff
@@ -0,0 +1,126 @@
+    }
+    
+    func showLayerContents(_ barButtonItem: UIBarButtonItem?) {
+        var layerContentsVC: LayerContentsViewController
```
This doesn't need to be mutable.

```suggestion
        let layerContentsVC: LayerContentsViewController
```

philium commented on data-collection/data-collection/View Controllers/Map View Controller/MapViewController+Extras.swift:
```diff
@@ -0,0 +1,126 @@
+    }
+    
+    func showBookmarks(_ barButtonItem: UIBarButtonItem?) {
+        var bookmarksVC: BookmarksViewController
```
This doesn't need to be mutable.

```suggestion
        let bookmarksVC: BookmarksViewController
```

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.swift:
```diff
@@ -0,0 +1,220 @@
+    var layerIndentationLevel: Int = 0 {
+        didSet {
+            // Set constraint constant; use +1 to account for initial indentation.
+            indentationConstraint.constant = CGFloat(layerIndentationLevel + 1) * indentationConstant
```
There's no limit on the sublayer count except what's supported by layers/web map.  I'm Ok without a limit and leaving it to the app/map developer, but I'll make the change it you want.

@philium, thoughts?

philium replied:
@mhdostal I think I like the idea of leaving it up to the app developer to handle. Any limit we impose would just be arbitrary and could prevent legitimate use cases, like iPad-only apps that take advantage of the whole screen.

esreli replied:
Sorry, I should have been more clear. The suggestion was not to limit the number of sublayers supported by the tool but rather limiting sublayer indentation at a certain depth.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.swift:
```diff
@@ -0,0 +1,220 @@
+import UIKit
+import ArcGIS
+
+let indentationConstant: CGFloat = 10.0
```
I move the cell reuse identifiers, but `indentationConstant` is used by both `UITableViewCells`.

philium replied:
I don't see a problem with global variables as long as they're `private`. Could you please privatize this?

philium replied:
```suggestion
private let indentationConstant: CGFloat = 10.0
```

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,480 @@
+/// The `LayerContentsViewController` can be styled to that of a legend, table of contents or some custom derivative.
+/// - Since: 100.8.0
+public class LayerContentsViewController: UIViewController {
+    /// Provide an out of the box TOC configuration.
+    internal struct TableOfContents: LayerContentsConfiguration {
+        var layersStyle: ConfigurationStyle = .allLayers
+        var allowToggleVisibility: Bool = true
+        var allowLayersAccordion: Bool = true
+        var showSymbology: Bool = true
+        var respectInitialLayerOrder: Bool = false
+        var respectShowInLegend: Bool = false
+        var showRowSeparator: Bool = true
+        var title: String = "Table of Contents"
+    }
+    
+    /// Provide an out of the box Legend configuration.
+    internal struct Legend: LayerContentsConfiguration {
+        var layersStyle: ConfigurationStyle = .visibleLayersAtScale
+        var allowToggleVisibility: Bool = false
+        var allowLayersAccordion: Bool = false
+        var showSymbology: Bool = true
+        var respectInitialLayerOrder: Bool = false
+        var respectShowInLegend: Bool = true
+        var showRowSeparator: Bool = false
+        var title: String = "Legend"
+    }
```
The superclass (LayerContentsViewController) uses the `TableOfContents` struct as it's default; otherwise, I would agree.

mhdostal commented on data-collection/data-collection/View Controllers/Map View Controller/MapViewController+Extras.swift:
```diff
@@ -0,0 +1,126 @@
+                })
+            }
+            
+            action.addAction(extraAction)
```
I'm inclined to not include images, since they are not explicitly supported by Apple.

mhdostal commented:
@esreli, @philium I've addressed all the comments.  Unless I missed something, the only outstanding item item is:

- potentially show a title and "Done" button when displaying either the TOC or Bookmarks VC.

mhdostal commented:
@philium - can you give this a final(?) look?

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,486 @@
+    }
+
+    var kind: Kind = .layer
+    var content: AnyObject?
```
Content.content is a bit odd, is there a better name for the class or this variable?

mhdostal replied:
`object`?  So `Content.object` or `DisplayedContent.object`?

rolson replied:
object, value - those might be better.

philium replied:
I like `object` better. Another idea is to combine this property with `kind` using an enum with associated values:

```swift
enum Object {
    case layer(AGSLayer)
    case sublayer(AGSLayerContent)
    case legendInfo(AGSLegendInfo)
}

mhdostal replied:
@philium - would you use the enum in place of both `kind` and `content`?  If so, how would that work?

mhdostal replied:
I've updated the code to use the `enum Object` @philium suggested.  It cleaned up the `tableView:cellForRowAt:` method a bit.

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,486 @@
+    }
+
+    var kind: Kind = .layer
+    var content: AnyObject?
```
instead of AnyObject, have you considered making Content generic?

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,486 @@
+    var allowToggleVisibility: Bool = false
+    var isVisibilityToggleOn: Bool = false
+    var isVisibleAtScale: Bool = true
+    var parents = [Content]()
```
does this create retain cycles?

mhdostal replied:
No, because the parent doesn't have a reference to the children.

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,486 @@
+    public var dataSource: LayerContentsDataSource? = nil {
+        didSet {
+            // Add an observer to handle changes to the dataSource.layerContents.
+            dataSourceObservation = dataSource?.observe(\.layerContents) { [weak self] (_, _) in
```
if this is just one to one communication, then a delegate method is clearer than KVO, IMO

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,486 @@
+    
+    // MARK: - Utility
+    
+    /// Returns a unique UINT for each object. Used because AGSLayerContent is not hashable
```
UInt

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,486 @@
+    
+    /// Returns a unique UINT for each object. Used because AGSLayerContent is not hashable
+    /// and we need to use it as the key in our dictionary of legendInfo arrays.
+    private static func objectIdentifierFor(_ obj: AnyObject) -> UInt {
```
I think you could instead use Hashable where you have AnyObject and use the hash for holding values in dictionaries.

mhdostal replied:
I tried `Hashable` briefly, but couldn't figure it out.  I did make the change to `objectId` -> `objectIdentifier` to be clearer.

philium replied:
Here's how you could conform `Kind` to `Hashable`:

```swift
extension LayerContentsRowConfiguration.Kind: Hashable {
    private var wrappedObject: AnyObject {
        switch self {
        case .layer(let layer):
            return layer
        case .sublayer(let layerContents):
            return layerContents
        case .legendInfo(let legendInfo):
            return legendInfo
        }
    }
    
    static func == (lhs: LayerContentsRowConfiguration.Kind, rhs: LayerContentsRowConfiguration.Kind) -> Bool {
        lhs.wrappedObject === rhs.wrappedObject
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(ObjectIdentifier(wrappedObject))
    }
}
```

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,486 @@
+}
+
+private extension AGSLayerContent {
+    func objectId() -> UInt {
```
If you don't go with the hashable approach, then I would call this `objectIdentifier` so it's not conflated with Feature objectIDs

rolson commented on data-collection/data-collection/View Controllers/Map View Controller/MapViewController+Extras.swift:
```diff
@@ -0,0 +1,149 @@
+import ArcGISToolkit
+
+/// Defines the Extras.
+public enum Extras : CaseIterable {
```
Is there a more descriptive name that we can use than Extras?

mhdostal replied:
@rolson Do either `MapViewExtras` or `MapViewControllerExtras` work?

cc: @esreli

rolson replied:
Sure, the latter seems fine.

mhdostal commented on data-collection/data-collection/Assets.xcassets/caret-down.imageset/Contents.json:
```diff
@@ -0,0 +1,26 @@
+  "images" : [
+    {
+      "idiom" : "universal",
+      "filename" : "caret-down@1x.png",
```
Those are from the Calcite Fonts [repo](https://github.com/ArcGIS/calcite-fonts) on GitHub.  Nathan created a script to generate iOS Assets from them.  I'll talk to @esreli ...

esreli replied:
My initial thought was the same as @rolson 's, the graphic looks a bit dated. Unfortunately, we can't use SF Symbols yet because it requires a min target of 13.0 and Data Collection supports the same min target as the Runtime.

Also, I really like the notion of promoting Esri assets as a foundation to OSA design . I did find another calcite graphic that might work for us, the "chevron". What do you think?

![chevron-right@3x](https://user-images.githubusercontent.com/33433113/81331729-e0b88e00-9056-11ea-9aef-e5a85ff29ccf.png)

rolson replied:
That could work better. Can we see both in context and make a visual comparison?

mhdostal replied:
Caret - 
![Caret](https://user-images.githubusercontent.com/3998072/81333232-e2d81800-9069-11ea-8761-a6c48a1e2235.png)

Chevron - 
![Chevron](https://user-images.githubusercontent.com/3998072/81333229-e23f8180-9069-11ea-9cbb-eacdbfef3420.png)

esreli replied:
I like that it feels less heavy.

rolson replied:
I like the Chevron much better

mhdostal replied:
Ok, chevron it is.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsDataSource.swift:
```diff
@@ -0,0 +1,123 @@
+    /// layer is at the top and drawn last).
+    /// - Since: 100.8.0
+    @objc
+    public dynamic private(set) var layerContents = [AGSLayerContent]()
```
Yes, KVO support. That way the TOC knows when the list of layers gets updated.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsDataSource.swift:
```diff
@@ -0,0 +1,123 @@
+            mapView.map?.load { [weak self] (error) in
+                guard let self = self else { return }
+                if let error = error {
+                    print("Error loading map: \(error)")
```
Not strictly necessary; you are correct, that the app should handle that type of error.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,486 @@
+
+/// Defines how to display layers in the table.
+/// - Since: 100.8.0
+internal class Content {
```
How about `DisplayedContent`?  Any other suggestions?

rolson replied:
I'm looking at the doc for this and trying to come up with names. DisplayedContent doesn't seem better because if I'm looking at it from an app perspective, that is ambiguous. If it was nested under LayerContentsViewController that's better. But maybe only necessary if a better name can't be thought of.

rolson replied:
Is this a "Row"? I didn't go very deep in the review so I'm not quite sure how this is used.

mhdostal replied:
Yes, it is a Row in the Table.  `RowContent`?

rolson replied:
The problem I'm looking to solve is that the fully qualified name of this type - if I look at it with all the other types in the app, do I know what it does? RowContent is better, but still lacks the "LayerContentsViewController" context IMO.

mhdostal replied:
Made the change to `LayerContentsRowConfiguration`.  It's explicit and states exactly what it is:  the configuration for a layer contents row in the table view.

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,486 @@
+public protocol LayerContentsConfiguration {
+    /// Specifies the `ConfigurationStyle` applied to the table.
+    /// - Since: 100.8.0
+    var layersStyle: ConfigurationStyle { get }
```
ConfigurationStyle should probably be nested under `LayerContentsConfiguration`. Without that, it's ambiguous

mhdostal replied:
`LayerContentsConfiguration` is a protocol and the compiler complains when I try to put the enum there.  I changed the name to `LayerContentsConfigurationStyle`.

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.swift:
```diff
@@ -0,0 +1,227 @@
+                    legendInfoCell.symbolImage = nil
+                    
+                    // We don't have a swatch for the given symbol, so create the swatch.
+                    symbol.createSwatch(completion: { [weak self] (swatch, _) -> Void in
```
I would have thought that the cell itself would call createSwatch and handle that internally, instead of a parent object doing it and setting it on the cell.

The way I've done this in the past is within the cell have a property for the cells' current "Value" or "Content", and on didSet call createSwatch. Within the completion of createSwatch check that your current value is the same as the value that you created the swatch for. Then there is no index matching, looking up indexPath for, etc. 

Also it compartmentalizes the fact that the cell does it's own operations.

However, I'm not dogmatic about this changing, but wanted to offer it as a suggestion.

You would probably need to pass a pointer to the cache when you create the cell so that the cell has access to the swatch cache.

philium replied:
@rolson See previous discussion [here](https://github.com/Esri/data-collection-ios/pull/230#discussion_r412732348), [here](https://github.com/Esri/data-collection-ios/pull/230#discussion_r412746326), [here](https://github.com/Esri/data-collection-ios/pull/230), and [here](https://github.com/Esri/data-collection-ios/pull/230#discussion_r414875437).

rolson replied:
ok - well it's fine the way it is then. Personally I think it's cleaner for the cell to hold the model and set it's UI. There are ways to solve the problem of it holding onto the model for too long I think.

mhdostal commented:
@philium, @rolson I made the changes requested, please have another look.

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsDataSource.swift:
```diff
@@ -56,10 +60,11 @@ public class LayerContentsDataSource: NSObject {
     
     private var mapOrSceneObservation: NSKeyValueObservation?
+    
+    public var delegate: LayerContentsDataSourceDelegate?
```
delegates should be weak usually to avoid retain cycle

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsDataSource.swift:
```diff
@@ -102,11 +109,13 @@ public class LayerContentsDataSource: NSObject {
             // Insert any base layers at the beginning of the `layerContents` array.
             if let baseLayers = basemap.baseLayers as? [AGSLayerContent] {
                 self.layerContents.insert(contentsOf: baseLayers, at: 0)
+                self.delegate?.layerContentsDidChange(self.layerContents)
```
You don't need this line everywhere you change layerContents. All you need to do is put this line in `didSet` on the layerContents property. Even if you do an `insert(...)`, `didSet` will get called IIRC.

mhdostal replied:
I've incorporated both suggestions.  I needed to pass a completion handler into the `appendBasemap` function because it needs to load the baseman and so is async.

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -69,7 +69,7 @@ public protocol LayerContentsConfiguration {
     override public init() {
         super.init()
-        config = LayerContentsViewController.Legend()
+        configuration = LayerContentsViewController.Legend()
```
probably can just say `.Legend()`

mhdostal replied:
I had tried that but the compiler won't let me, since `Legend` is not a property of `LayerContentsConfiguration` (which is the type for `configuration`).

rolson commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -82,26 +82,26 @@ public class LegendViewController: LayerContentsViewController {
     override public init() {
         super.init()
-        config = LayerContentsViewController.TableOfContents()
+        configuration = LayerContentsViewController.TableOfContents()
```
can probably just say `.TableOfContents()`

mhdostal replied:
same as above.

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsTableViewController.swift:
```diff
@@ -0,0 +1,231 @@
+        let cell: UITableViewCell
+        let rowItem: LayerContentsRowConfiguration = visibleConfigurations[indexPath.row]
+        switch rowItem.kind {
+        case .layer:
+            // rowItem is a layer or a sublayer which implements AGSLayerContent
+            let layerCell = (tableView.dequeueReusableCell(withIdentifier: LayerContentsTableViewController.layerCellReuseIdentifier) as! LayerCell)
+            cell = layerCell
+            setupLayerCell(layerCell, indexPath, rowItem)
+        case .sublayer:
+            // rowItem is a layer or a sublayer which implements AGSLayerContent
+            let layerCell = (tableView.dequeueReusableCell(withIdentifier: LayerContentsTableViewController.sublayerCellReuseIdentifier) as! LayerCell)
+            cell = layerCell
+            setupLayerCell(layerCell, indexPath, rowItem)
```
There are two cells, one for layers and one for sublayers, so that the client can customize the look for each in the storyboards.  The prototype cells currently differ in how the "name" label is formatted.

See the discussion [here](https://github.com/Esri/data-collection-ios/pull/230#discussion_r413069403).

philium commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,528 @@
+            parentArray.forEach { parent in
+                rowConfigurations.forEach { potentialParentContent in
+                    // Search all contents to see if it's a match.
+                    if potentialParentContent.kind.matches(object: parent) {
```
Since `Kind` conforms to `Equatable`, this check could just test for equality:

```suggestion
                    if potentialParentContent.kind == .sublayer(parent) {
```

And then `matches(object:)` wouldn't be needed.

philium commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,528 @@
+                
+                if (configuration.layersStyle == .visibleLayersAtScale && showAtScale) || configuration.layersStyle == .allLayers {
+                    let internalLegendInfos = legendInfos[layerContent.objectIdentifier()] ?? []
+                    let rowConfiguration = LayerContentsRowConfiguration(LayerContentsRowConfiguration.Kind.sublayer(subLayerContent),
```
```swift
                    let rowConfiguration = LayerContentsRowConfiguration(.sublayer(subLayerContent),
```

philium commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,528 @@
+            if let internalLegendInfos = legendInfos[layerContent.objectIdentifier()] {
+                let showAtScale = parentShowAtScale && shouldShowAtScale(layerContent)
+                let contentArray = internalLegendInfos.map { legendInfo -> LayerContentsRowConfiguration in
+                    let rowConfiguration = LayerContentsRowConfiguration(LayerContentsRowConfiguration.Kind.legendInfo(legendInfo),
```
```swift
                    let rowConfiguration = LayerContentsRowConfiguration(.legendInfo(legendInfo),
```

mhdostal commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,528 @@
+    private var layerContentsTableViewController: LayerContentsTableViewController?
+    
+    // Dictionary of legend infos; keys are AGSLayerContent objectIdentifier values.
+    private var legendInfos = [UInt: [AGSLegendInfo]]()
```
That's an excellent idea, thanks!  Simplifies it a bunch.

philium commented on data-collection/data-collection/View Controllers/Layer Contents View Controllers/LayerContentsViewController.swift:
```diff
@@ -0,0 +1,517 @@
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(ObjectIdentifier(wrappedObject))
+    }
+    
+    /// Compares the private `Kind.wrappedObject` to a given `object`.
+    /// - Parameter object: The object to compare to `Kind.wrappedObject`
+    /// - Returns: `true` if it matches, `false` otherwise.
+    func matches(object: AnyObject) -> Bool {
+        var kind: LayerContentsRowConfiguration.Kind?
+        switch object {
+        case let layer as AGSLayer:
+            kind = .layer(layer)
+        case let layerContent as AGSLayerContent:
+            kind = .sublayer(layerContent)
+        case let legendInfo as AGSLegendInfo:
+            kind = .legendInfo(legendInfo)
+        default:
+            kind = nil
+        }
+
+        return self == kind
+    }
```
This method is no longer being used. Could you please remove it?

```suggestion
```

mhdostal commented:
@philium - Awesome, thank you for all the help!

philium commented:
@mhdostal You're welcome! I'm happy to help.

mhdostal commented:
@rolson Thanks!
